File: frontend/src/types/graph.ts
export type NodeType =
  | 'repository'
  | 'directory'
  | 'file'
  | 'module'
  | 'namespace'
  | 'class'
  | 'interface'
  | 'trait'
  | 'enum'
  | 'function'
  | 'method';

export type EdgeType =
  | 'CONTAINS'
  | 'DEFINED_IN'
  | 'IMPORTS'
  | 'USES_MODULE'
  | 'CALLS'
  | 'REFERENCES'
  | 'EXTENDS'
  | 'IMPLEMENTS'
  | 'USES_TRAIT'
  | 'OVERRIDES'
  | 'DEPENDS_ON'
  | 'POSSIBLE_CALLS'
  | 'POSSIBLE_REFERENCES';

export type ProvenanceTag =
  | 'parser_fact'
  | 'reference_index_fact'
  | 'graph_algorithm_inference'
  | 'model_assisted_inference';

export type ExternalKind = 'builtin' | 'stdlib' | 'third_party' | 'unknown';

export interface GraphNodeProperties {
  relative_path?: string;
  module_name?: string;
  external_kind?: ExternalKind;
  dependency_manifest?: string;
  dependency_name?: string;
  dependency_version?: string | null;
  ecosystem?: string;
  [key: string]: unknown;
}

export interface GraphEdgeProperties {
  refined_from?: string;
  layer?: string;
  [key: string]: unknown;
}

export interface GraphNode {
  id: string;
  type: NodeType;
  label: string;
  language: string;
  provenance: ProvenanceTag;
  confidence: number;
  file_path?: string | null;
  line_start?: number | null;
  line_end?: number | null;
  qualified_name?: string | null;
  docstring?: string | null;
  signature?: string | null;
  decorators?: string[];
  is_async?: boolean;
  is_entry_point?: boolean;
  hotspot_score?: number;
  ai_summary?: string | null;
  properties?: GraphNodeProperties;
}

export interface GraphEdge {
  id: string;
  type: EdgeType;
  source_id: string;
  target_id: string;
  language: string;
  confidence: number;
  evidence?: string | null;
  file_path?: string | null;
  line_number?: number | null;
  provenance: ProvenanceTag;
  provenance_detail?: string | null;
  model_name?: string | null;
  properties?: GraphEdgeProperties;
}

export interface GraphStats {
  node_count: number;
  edge_count: number;
  file_count: number;
  class_count: number;
  function_count: number;
  method_count: number;
  unresolved_calls: number;
  model_inferred_edges: number;
}

export interface GraphResponse {
  nodes: GraphNode[];
  edges: GraphEdge[];
  stats: GraphStats;
}

export interface GraphFilters {
  minConfidence: number;
  q: string;
  includeParserFacts: boolean;
  includeReferenceFacts: boolean;
  includeGraphInference: boolean;
  includeModelInference: boolean;
  hideNativeLibraryNodes: boolean;
  hideThirdPartyDependencyNodes: boolean;
  showProvenanceBadges: boolean;
  groupByModules: boolean;
}

File: backend/models/graph.py
from __future__ import annotations
from enum import Enum
from typing import Optional, Any
from pydantic import BaseModel


class NodeType(str, Enum):
    REPOSITORY = "repository"
    DIRECTORY = "directory"
    FILE = "file"
    MODULE = "module"
    NAMESPACE = "namespace"
    CLASS = "class"
    INTERFACE = "interface"
    TRAIT = "trait"
    ENUM = "enum"
    FUNCTION = "function"
    METHOD = "method"


class EdgeType(str, Enum):
    CONTAINS = "CONTAINS"
    DEFINED_IN = "DEFINED_IN"
    IMPORTS = "IMPORTS"
    USES_MODULE = "USES_MODULE"
    CALLS = "CALLS"
    REFERENCES = "REFERENCES"
    EXTENDS = "EXTENDS"
    IMPLEMENTS = "IMPLEMENTS"
    USES_TRAIT = "USES_TRAIT"
    OVERRIDES = "OVERRIDES"
    DEPENDS_ON = "DEPENDS_ON"
    POSSIBLE_CALLS = "POSSIBLE_CALLS"
    POSSIBLE_REFERENCES = "POSSIBLE_REFERENCES"


class ProvenanceTag(str, Enum):
    PARSER_FACT = "parser_fact"
    REFERENCE_INDEX_FACT = "reference_index_fact"
    GRAPH_ALGORITHM_INFERENCE = "graph_algorithm_inference"
    MODEL_ASSISTED_INFERENCE = "model_assisted_inference"


class Language(str, Enum):
    PYTHON = "python"
    PHP = "php"
    MIXED = "mixed"
    UNKNOWN = "unknown"


class GraphNode(BaseModel):
    id: str
    type: NodeType
    label: str
    language: Language = Language.UNKNOWN
    provenance: ProvenanceTag = ProvenanceTag.PARSER_FACT
    confidence: float = 1.0
    file_path: Optional[str] = None
    line_start: Optional[int] = None
    line_end: Optional[int] = None
    qualified_name: Optional[str] = None
    docstring: Optional[str] = None
    signature: Optional[str] = None
    decorators: list[str] = []
    is_async: bool = False
    is_entry_point: bool = False
    hotspot_score: float = 0.0
    ai_summary: Optional[str] = None
    properties: dict[str, Any] = {}


class GraphEdge(BaseModel):
    id: str
    type: EdgeType
    source_id: str
    target_id: str
    language: Language = Language.UNKNOWN
    confidence: float = 1.0
    evidence: Optional[str] = None
    file_path: Optional[str] = None
    line_number: Optional[int] = None
    provenance: ProvenanceTag = ProvenanceTag.PARSER_FACT
    provenance_detail: Optional[str] = None
    model_name: Optional[str] = None
    properties: dict[str, Any] = {}


class GraphStats(BaseModel):
    node_count: int = 0
    edge_count: int = 0
    file_count: int = 0
    class_count: int = 0
    function_count: int = 0
    method_count: int = 0
    unresolved_calls: int = 0
    model_inferred_edges: int = 0
